rsky-pds/src/repo/actor_store.rs

```diff
@@ -0,0 +1,369 @@
+                                Some(now.to_string()),
+                            )
+                            .await?
+                    }
+                    PreparedWrite::Delete(write) => self.record.delete_record(write.uri).await?,
+                })
+            })
+            .collect::<Vec<_>>()
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+        Ok(())
+    }
+
+    pub async fn destroy(&mut self) -> Result<()> {
+        use crate::schema::pds::blob::dsl as BlobSchema;
+        let conn = &mut establish_connection()?;
+
+        let blob_rows: Vec<String> = BlobSchema::blob
+            .filter(BlobSchema::did.eq(&self.did))
+            .select(BlobSchema::cid)
+            .get_results(conn)?;
+        let cids = blob_rows
+            .into_iter()
+            .map(|row| Ok(Cid::from_str(&row)?))
+            .collect::<Result<Vec<Cid>>>()?;
+        let _ = stream::iter(cids.chunks(500))
+            .then(|chunk| async {
+                Ok::<(), anyhow::Error>(self.blob.blobstore.delete_many(chunk.to_vec()).await?)
+            })
+            .collect::<Vec<_>>()
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+        Ok(())
+    }
+
+    // @TODO: Use AtUri
+    pub async fn get_duplicate_record_cids(
+        &self,
+        cids: Vec<Cid>,
+        touched_uris: Vec<String>,
+    ) -> Result<Vec<Cid>> {
+        if touched_uris.len() == 0 || cids.len() == 0 {
+            return Ok(vec![]);
+        }
+        use crate::schema::pds::record::dsl as RecordSchema;
+        let conn = &mut establish_connection()?;
+
+        let cid_strs: Vec<String> = cids.into_iter().map(|c| c.to_string()).collect();
+        let res: Vec<String> = RecordSchema::record
+            .filter(RecordSchema::did.eq(&self.did))
+            .filter(RecordSchema::cid.eq_any(cid_strs))
+            .filter(RecordSchema::uri.ne_all(touched_uris))
+            .select(RecordSchema::cid)
+            .get_results(conn)?;
+        Ok(res
+            .into_iter()
+            .map(|row| Cid::from_str(&row).map_err(|error| anyhow::Error::new(error)))
+            .collect::<Result<Vec<Cid>>>()?)
+    }
+
+    pub async fn reserve_keypair(&self, did: Option<&str>) -> Result<String> {
+        if let Some(did) = did {
+            assert_safe_path_part(&did);
+            let key_loc = Path::new(&self.reserved_key_dir).join(did);
+            let key = load_key(key_loc);
+            if key.is_ok() {
+                return Ok(key?.did()?);
+            }
+        }
+        let keypair = Secp256k1Keypair::create(Some(Secp256k1KeypairOptions {
+            exportable: Some(true),
+        }))
+        .await?;
+        let key_did = keypair.did()?;
+        let key_loc = Path::new(&self.reserved_key_dir).join(&key_did);
+        fs::create_dir_all(self.reserved_key_dir.clone())?;
+        fs::write(key_loc, keypair.export()?)?;
+        Ok(key_did)
+    }
```

Comment:
[ts impl for reference](https://github.com/bluesky-social/atproto/blob/12dcdb668c8ec0f8a89689c326ab3e9dbc6d2f3c/packages/pds/src/actor-store/index.ts#L173)

would appreciate a second set of eyes here - i want to make sure that i understand the conditional flow of this function.  it seems to me like if we have a `did` that we pass in, then we'd want to load the key at that path and return the `did` of that key.  otherwise, we go and create a new key, save it to a location, and return that new `did`.  

if that's correct, this should probably be rewritten as a `match did { Some(did) => {...}, None => {...} }` expression.  right now, there's the case where we do pass a `did`, but loading the file fails, and we create a new key instead (i.e. the `if` condition on line 330 is false, `! key.is_ok()`).

but i'm just not too sure i have the logic right here.